keps/sig-api-machinery/storage-migration.md

```diff
@@ -141,29 +149,57 @@ Read the [workflow section][] to understand how the API is used.
 ```golang
 type StorageVersionMigration struct {
   metav1.TypeMeta
-  // metadata.name should be "<resource>.<group>" of the resource being
-  // migrated.
-  // Because resouces can be served at different groups, in the future version,
-  // metadata.name can be hash(etcd instance, etcd path prefix).
+  // At the alpha stage, metadata.name must be "<resource>.<group>" of the
+  // resource being migrated.
+  // We require this name format to avoid duplicate `migration` CRs for the same
+  // resource type. Otherwise we will need to build complicated system to make
+  // sure the latest `migration` CR is honored. 
+  // Using "<resource>.<group>" isn't entirely safe, for two reasons:
+  // a. resources share key spaces can be served at different groups
+  // b. groupX/v1/foo and groupX/v5/foo may not share key spaces and thus
+  // require two `migration` CRs, though Kubernetes don't have such cases yet. 
+  // To address the two concerns, in the beta version, the name need to be
+  // tied to the key spaces, e.g., a hash of "etcd instance"+"etcd path prefix".
   metav1.ObjectMeta
   Spec StorageVersionMigrationSpec
+  // If status is empty, it means the `migration` is just created. The migrator
+  // controller hasn't launched the **migration worker** `job` for it yet.
   Status StorageVersionMigrationStatus
 }
 
+// Note that the spec only contains an immutable field in the alpha version. To
+// request anohter round of migration, client needs to delete and recreate the
+// `migration` CR.
 type StorageVersionMigrationSpec {
   // Resource is the resource that is being migrated. The migrator sends
-  // requests to the endpoint serving the Resource.
+  // requests to the endpoint tied to the Resource.
   // Immutable.
   Resource GroupVersionResource
 }
 
+type MigrationPhase string
+
+const (
+  // MigrationPending indicates that the migration is requested, but no 
+  // migrator job has been created to fulfill the request.
+  MigrationPending MigrationPhase = "Pending"
+  // MigrationRunning indicates that a migrator job is running.
+  MigrationRunning MigrationPhase = "Running"
+  // MigrationSucceed indicates that the migration has completed successfully.
+  MigrationSucceeded MigrationPhase = "Succeeded"
+  // MigrationSucceed indicates that the migration has failed.
+  MigrationRunning MigrationPhase = "Running"
+)
+
 type StorageVersionMigrationStatus {
-  // Completed means the migration of this resource has completed.
-  Completed bool
+  // Phase indicates where the migration is in the life-cycle. Do not treat it
+  // as a state machine, because new phases can be added.
```

Comment:
They can?

I think this is an impossible ask from clients.

There is confusion about this, but I think the latest guidance is still that extensible things like this should be represented as conditions.

But does this need to be extensible?

Reply:
Converted to the conditions pattern.